### DISCUSSION TITLE
build project is cmake, How to clean the clangd cache?

### DISCUSSION THREAD

**Discussion by Hebe970217** [clangd/clangd#1001] [Q&A]
build project is cmake, How to clean the clangd cache? 
my steps:
1. to used "cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=1" generate "compile_commands.json" and copy the file to project root path.
2. to start clangd, I watch the generated ".cache" folder from project root path.（Index file）
3. remove ".cache" and "compile_commands.json" for project root path
4. restart clangd. cache files will still be generated（Index file）

In addition, I also want to know how "-- background index" affects the above steps

**Comment by sam-mccall** [comment]
`.cache` stores the background index for the project.

You can clean it by removing `.cache`, and it will be regenerated next time indexing happens.

If you don't want background indexing, you can turn it off with `--background-index=0`.

If you want background indexing but don't want it stored in that directory, we don't currently have an option for that. #184

**Reply by Hebe970217** [reply]
compile_ commands. json must be under the project file to start the Clangd index operation.
I deleted this json file. But The cache folder is still generated.
I don't quite understand.

**Reply by Hebe970217** [reply]
Before copying, compile_ commands.json the file is in the subdirectory of project root path, name is ".unioncode"

**Reply by sam-mccall** [reply]
Can you please provide the full folder structure? (absolute paths, where the source code is, the compilation database is, where the .cache dir is generated)

**Reply by Hebe970217** [reply]
show dir:
```base
funning@funning-PC:~/workspace/workspace/recode/gg/unioncode$ ls -al
total 88
drwxr-xr-x 12 funning funning  4096 Jan 27 17:28 .
drwxr-xr-x 11 funning funning  4096 Jan 24 18:40 ..
drwxr-xr-x  4 funning funning  4096 Jan 13 14:58 3rdparty
drwxr-x---  3 funning funning  4096 Jan 27 17:28 .cache
-rw-r--r--  1 funning funning  3317 Dec 31 13:31 CMakeLists.txt
-rw-r--r--  1 funning funning 26693 Jan 24 09:35 CMakeLists.txt.user
drwxr-xr-x  4 funning funning  4096 Jan 19 12:02 debian
drwxr-xr-x  3 funning funning  4096 Jan 25 17:19 docs
drwxr-xr-x  8 funning funning  4096 Jan 27 18:27 .git
-rw-r--r--  1 funning funning   869 Dec 31 13:31 .gitignore
-rw-r--r--  1 funning funning     1 Nov 12 17:45 README.md
drwxr-xr-x 12 funning funning  4096 Dec 29 15:31 src
drwxr-xr-x  2 funning funning  4096 Nov 26 18:04 tests
drwxr-xr-x  2 funning funning  4096 Nov 12 17:53 translations
drwxr-xr-x  6 funning funning  4096 Jan 27 17:28 .unioncode

funning@funning-PC:~/workspace/workspace/recode/gg/unioncode$ ls -al .unioncode/
total 632
drwxr-xr-x  6 funning funning   4096 Jan 27 17:28 .
drwxr-xr-x 12 funning funning   4096 Jan 27 17:28 ..
drwxr-xr-x  5 funning funning   4096 Jan 27 17:28 3rdparty
-rw-r--r--  1 funning funning  24707 Jan 27 17:28 CMakeCache.txt
drwxr-xr-x  4 funning funning   4096 Jan 27 17:28 CMakeFiles
-rw-r--r--  1 funning funning   1973 Jan 27 17:28 cmake_install.cmake
-rw-r--r--  1 funning funning 553031 Jan 27 17:28 compile_commands.json
-rw-r--r--  1 funning funning    382 Jan 27 17:28 CTestTestfile.cmake
-rw-r--r--  1 funning funning  26854 Jan 27 17:28 Makefile
drwxr-xr-x 11 funning funning   4096 Jan 27 17:28 src
drwxr-xr-x  3 funning funning   4096 Jan 27 17:28 tests
```
"~/workspace/workspace/recode/gg/unioncode" is project root path.

"~/workspace/workspace/recode/gg/unioncode/.unioncode" is build dir.

"compile_commands.json" in build dir

**Reply by Hebe970217** [reply]
I keep the above directory structure unchanged,
Delete After the .cache folder, Clangd will still be generated .cache

**Reply by Hebe970217** [reply]
The compile_commands.json file only exists in the subdirectory "~/workspace/workspace/recode/gg/unioncode/.unioncode" of the project root path.
Is this normal?

**Comment by sam-mccall** [comment]
Expected behavior:
 - you need to copy or link compile_commands.json from your build directory into your source directory for things to work. (Unless your build directory is named `build`)
 - `.cache` is created in your project directory, which is the source root containing `compile_commands.json` (or containing `build/compile_commands.json`)

Unexpected behavior:
 - if you do not copy `compile_commands.json` to your sources directory, clangd creates `.cache` there anyway.

Can you confirm you're seeing this?

**Reply by Hebe970217** [reply]
Thanks, I found the json file, and delete the json in name's build folder, it's working normal.
(Unless your build directory is named build) 
Maybe this rule should be marked. I read the document and didn't find it. that's Implicit rules?

I have new question, to setting compile_command.json set, Can I to do from"initialization" request (protocol) ?

**Reply by sam-mccall** [reply]
> Maybe this rule should be marked. I read the document and didn't find it. that's Implicit rules?

Good point, sorry! I added it to the docs in https://github.com/llvm/clangd-www/commit/8df3bda9b3c3eb4e52a7310970176eb3461dbdda

> I have new question, to setting compile_command.json set, Can I to do from"initialization" request (protocol) ?

You can specify the database to use using an LSP extension in the initialize request, yes: https://clangd.llvm.org/extensions#compilation-commands

`compilationDatabasePath` lets you specify the directory containing `compile_commands.json`

`compilationDatabaseChanges`/`fallbackFlags` lets you specify flags directly over LSP, without having a `compile_commands.json` file on disk at all. (I can't remember exactly how this interacts with background indexing, I can dig this up if you need to know)

**Reply by Hebe970217** [reply]
I may not need to know how to index the database. Thanks again. 

I'm on another question，see the questions raised by others about "how to set include path to Clangd", it seems that there is no conclusion.
My project is normally opened in other IDE, but after starting Clangd with "complex_command.json", I find that the relevant header file cannot be found. 
This message comes from the "textdocument / publishdiagnostics" event of the server

**Reply by Hebe970217** [reply]
The Error message:
```base
In includeed file: 'stddef.h' file not found
```
the stddef.h is  C++ standard library, my system is linux.
found it path from "/usr/include/linux/stddef.h".

**Reply by Hebe970217** [reply]
I search a cmake operations "-isystem"，can setting project include head from system?
in addition, How can I setting it from clangd operation?

**Comment by mellery451** [comment]
sorry to piggy-back here, but I find that I have TWO cache directories:

     <repo-root>/.cache/clangd   (<repo-root is where the compile command DB lives and where I generally launch my editor etc)
     $HOME/.cache/clangd

Both locations seem to have somewhat recent files - what's the difference? If I want to clean the cache, should I remove both location?

**Reply by HighCommander4** [reply]
I believe the second one is a fallback for storing information about files for which a `compile_commands.json` file could not be found.
